### DISCUSSION TITLE
Can't compile on macos

### DISCUSSION THREAD

**Discussion by fppsvz** [deepmodeling/deepmd-kit#8] [Q&A] [answered]
Hello to everybody.
I'm writing cause I'm trying to compile DeepMD on my MacBookPro (running macos 10.13) but with scarce results. I've been able to install everything up to the actual DeepMD.
In particular, following strictly the steps reported here in the guide, I arrived up to the execution of cmake without any issues:
`cmake -DXDRFILE_ROOT=$xdrfile_root -DTENSORFLOW_ROOT=$tensorflow_root -DCMAKE_INSTALL_PREFIX=$deepmd_root -DCMAKE_C_COMPILER=/usr/local/Cellar/gcc/8.2.0/bin/gcc-8 -DCMAKE_CXX_COMPILER=/usr/local/Cellar/gcc/8.2.0/bin/g++-8 ..`  

Here I specified the gnu compilers gcc cause I thought it could have been an issue with clang, but still without luck. I installed gcc8.2 using brew.

After I call `make` I get the following errors:

`[  1%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NNPAtomMap.cc.o
[  3%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NNPInter.cc.o
[  4%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NeighborList.cpp.o
[  6%] Building CXX object lib/CMakeFiles/deepmd.dir/src/SimulationRegion.cpp.o
[  8%] Linking CXX shared library libdeepmd.dylib
Undefined symbols for architecture x86_64:
  "tensorflow::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**)", refer`

[a lot of stuffs]
and finally: 

`ld: symbol(s) not found for architecture x86_64
collect2: error: ld returned 1 exit status
make[2]: *** [lib/libdeepmd.dylib] Error 1
make[1]: *** [lib/CMakeFiles/deepmd.dir/all] Error 2
make: *** [all] Error 2`.

I understand this is an issue in creating the dynamic shared libraries but I have no ideas on how to solve this. Does anybody can help me? That would be sooooo much appreciated.

Best regards,
Filippo

**Comment by frankhan91** [comment]
I would recommend using Docker to run DeepMD-kit on MacOS. 

Here are two related repos: https://github.com/frankhan91/deepmd-kit_docker and https://github.com/XiaoXX0522/deepmd-kit_docker

**Comment by yanqinz2** [comment]
> Hello to everybody.
> I'm writing cause I'm trying to compile DeepMD on my MacBookPro (running macos 10.13) but with scarce results. I've been able to install everything up to the actual DeepMD.
> In particular, following strictly the steps reported here in the guide, I arrived up to the execution of cmake without any issues:
> `cmake -DXDRFILE_ROOT=$xdrfile_root -DTENSORFLOW_ROOT=$tensorflow_root -DCMAKE_INSTALL_PREFIX=$deepmd_root -DCMAKE_C_COMPILER=/usr/local/Cellar/gcc/8.2.0/bin/gcc-8 -DCMAKE_CXX_COMPILER=/usr/local/Cellar/gcc/8.2.0/bin/g++-8 ..`
> 
> Here I specified the gnu compilers gcc cause I thought it could have been an issue with clang, but still without luck. I installed gcc8.2 using brew.
> 
> After I call `make` I get the following errors:
> 
> `[ 1%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NNPAtomMap.cc.o [ 3%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NNPInter.cc.o [ 4%] Building CXX object lib/CMakeFiles/deepmd.dir/src/NeighborList.cpp.o [ 6%] Building CXX object lib/CMakeFiles/deepmd.dir/src/SimulationRegion.cpp.o [ 8%] Linking CXX shared library libdeepmd.dylib Undefined symbols for architecture x86_64: "tensorflow::NewSession(tensorflow::SessionOptions const&, tensorflow::Session**)", refer`
> 
> [a lot of stuffs]
> and finally:
> 
> `ld: symbol(s) not found for architecture x86_64 collect2: error: ld returned 1 exit status make[2]: *** [lib/libdeepmd.dylib] Error 1 make[1]: *** [lib/CMakeFiles/deepmd.dir/all] Error 2 make: *** [all] Error 2`.
> 
> I understand this is an issue in creating the dynamic shared libraries but I have no ideas on how to solve this. Does anybody can help me? That would be sooooo much appreciated.
> 
> Best regards,
> Filippo

Hello Filippo,

I met exactly the same problem with yours. Have you found any solution yet? Thank you very much.

Best,
Yanqin

**Comment by fppsvz** [comment, marked as answer]
Hello Yanqin,
The solution suggested by frankhan91 solved my problem: use the Docker container.
Technically it is not installed directly on MacOS, but in a Docker container upon a CentOS 7 layer, but at the end of the day I'm running it on my MacBookPro so it's fine. Performance-wise I can't tell you if the docker layer is so detrimental. I usually run tests on my laptop, then the hardcore training I run them on a HPC.
Best,
Filippo

**Comment by yufengwhy** [comment]
conda install -c conda-forge deepmd-kit
